geo/src/algorithm/triangulate_spade.rs

```diff
@@ -0,0 +1,568 @@
+    /// ```
+    ///
+    /// Compared to the `constrained_outer_triangulation` it only includes the triangles
+    /// inside of the input geometry
+    fn constrained_triangulation(&'a self) -> TriangulationResult<Triangles<T>> {
+        self.constrained_outer_triangulation().map(|triangles| {
+            triangles
+                .into_iter()
+                .filter(|triangle| {
+                    let center = triangle.centroid();
+                    self.contains_point(center)
+                })
+                .collect::<Vec<_>>()
+        })
+    }
+}
+
+/// conversion from spade triangulation back to geo triangles
+fn triangulation_to_triangles<T, F>(triangulation: T) -> Triangles<F>
+where
+    T: Triangulation<Vertex = Point2<F>>,
+    F: SpadeTriangulationFloat,
+{
+    triangulation
+        .inner_faces()
+        .map(|face| face.positions())
+        .map(|points| points.map(|p| Coord::<F> { x: p.x, y: p.y }))
+        .map(Triangle::from)
+        .collect::<Vec<_>>()
+}
+
+// ========== Triangulation trait impls ============
+
+impl<'a, 'l, T, G> TriangulateSpade<'a, T> for G
+where
+    'a: 'l,
+    T: SpadeTriangulationFloat,
+    G: LinesIter<'l, Scalar = T> + CoordsIter<Scalar = T> + Contains<Point<T>>,
+{
+    fn get_points(&'a self) -> Vec<Coord<T>> {
+        self.coords_iter().collect::<Vec<_>>()
+    }
+
+    fn get_lines(&'a self) -> Vec<Line<T>> {
+        self.lines_iter().collect()
+    }
+
+    fn contains_point(&'a self, p: Point<T>) -> bool {
+        self.contains(&p)
+    }
+}
+
+// it would be cool to impl the trait for GS: AsRef<[G]> but I wasn't able to get this to compile
+// (yet)
+
+impl<'a, T, G> TriangulateSpade<'a, T> for Vec<G>
+where
+    T: SpadeTriangulationFloat,
+    G: TriangulateSpade<'a, T>,
+{
+    fn get_points(&'a self) -> Vec<Coord<T>> {
+        self.iter().flat_map(|g| g.get_points()).collect::<Vec<_>>()
+    }
+
+    fn get_lines(&'a self) -> Vec<Line<T>> {
+        self.iter().flat_map(|g| g.get_lines()).collect::<Vec<_>>()
+    }
+
+    fn contains_point(&'a self, p: Point<T>) -> bool {
+        self.iter().any(|g| g.contains_point(p))
+    }
+}
+
+impl<'a, T, G> TriangulateSpade<'a, T> for &[G]
+where
+    T: SpadeTriangulationFloat,
+    G: TriangulateSpade<'a, T>,
+{
+    fn get_points(&'a self) -> Vec<Coord<T>> {
+        self.iter().flat_map(|g| g.get_points()).collect::<Vec<_>>()
+    }
+
+    fn get_lines(&'a self) -> Vec<Line<T>> {
+        self.iter().flat_map(|g| g.get_lines()).collect::<Vec<_>>()
+    }
+
+    fn contains_point(&'a self, p: Point<T>) -> bool {
+        self.iter().any(|g| g.contains_point(p))
+    }
+}
+
+// ========== Triangulation trait impl helpers ============
+
+fn prepare_intersection_contraint<T: SpadeTriangulationFloat>(
+    mut lines: Vec<Line<T>>,
+    mut known_points: Vec<Coord<T>>,
+) -> Result<Vec<Line<T>>, TriangulationError> {
+    // safety net. We can't prove that the `while let` loop isn't going to run infinitely, so
```

Comment:
Could you document the conditions under which we might see an infinite loop? How did you arrive at the 1000 loops number?